rust/cubesql/cubesql/src/sql/postgres/service.rs

```diff
@@ -34,19 +34,40 @@ impl ProcessingLoop for PostgresServer {
 
         println!("🔗 Cube SQL (pg) is listening on {}", self.address);
 
-        let shim_cancellation_token = CancellationToken::new();
+        let fast_shutdown_interruptor = CancellationToken::new();
+        let semifast_shutdown_interruptor = CancellationToken::new();
 
         let mut joinset = tokio::task::JoinSet::new();
+        let mut active_shutdown_mode: Option<ShutdownMode> = None;
 
         loop {
             let mut stop_receiver = self.close_socket_rx.write().await;
             let (socket, _) = tokio::select! {
-                res = stop_receiver.changed() => {
-                    if res.is_err() || *stop_receiver.borrow() {
-                        trace!("[pg] Stopping processing_loop via channel");
-
-                        shim_cancellation_token.cancel();
-                        break;
+                _ = stop_receiver.changed() => {
+                    let mode = *stop_receiver.borrow();
+                    if mode > active_shutdown_mode {
+                        active_shutdown_mode = mode;
+                        match active_shutdown_mode {
+                            Some(ShutdownMode::Fast) => {
+                                trace!("[pg] Stopping processing_loop via channel, fast mode");
+
+                                fast_shutdown_interruptor.cancel();
+                                break;
+                            }
+                            Some(ShutdownMode::SemiFast) => {
+                                trace!("[pg] Stopping processing_loop via channel, semifast mode");
+
+                                semifast_shutdown_interruptor.cancel();
+                                break;
+                            }
+                            Some(ShutdownMode::Smart) => {
+                                trace!("[pg] Stopping processing_loop via interruptor, smart mode");
+                                break;
+                            }
+                            None => {
+                                panic!("Unreachable None case greater than some other value");
```

Comment:
WDYT about using `unreachable!` ?)

Reply:
n00b moment; replaced the two panics with unreachable.